### ISSUE TITLE
Questions about UTC<->TAI conversions

### ISSUE THREAD

**Issue by r-owen** [astropy/astropy#10055]
### Description

Leap second correction is applied gradually on the day of a leap second, leading to questions about conversion between UTC and TAI time. [EDIT by @taldcroft]

### Expected behavior

The leap second correction should be applied one second before the new day.

### Actual behavior

The leap second is applied gradually throughout the day.

### Steps to Reproduce

Here is a simplified example halfway through the day of the most recent leap second.  `tai_minus_utc1` is a variant of the code I usually run, which handles the leap second correctly. `tai_minus_utc2` is the code I would rather use, because it is simpler. See https://github.com/astropy/astropy/issues/10052 for another variant that is also simple, but fails in the same way as `tai_minus_utc2`:
```
import astropy.time

def tai_minus_utc1(ap_time):
    try:
        dt_utc = ap_time.utc.to_datetime()
        dt_tai = ap_time.tai.to_datetime()
    except ValueError:
        # datetime cannot have 60 in the seconds field; back off a second
        ap_time = ap_time - 1 * u.second
        dt_utc = ap_time.utc.to_datetime()
        dt_tai = ap_time.tai.to_datetime()
    return (dt_tai - dt_utc).total_seconds()

def tai_minus_utc2(ap_time):
    return (ap_time.tai.mjd - ap_time.utc.mjd) * 24 * 60 * 60

ap_time = astropy.time.Time("2016-12-31 12:00:00", scale="utc", format="iso")
print(f"tai_minus_utc1: {tai_minus_utc1(ap_time)}")
print(f"tai_minus_utc2: {tai_minus_utc2(ap_time)}")
```
produces:
tai_minus_utc1: 36.0  # the value I expected
tai_minus_utc2: 36.499994387850165  # wrong

### System Details

I see this on two different platforms:

Linux-4.19.76-linuxkit-x86_64-with-centos-7.7.1908-Core
Python 3.7.2 (default, Dec 29 2018, 06:19:36) 
[GCC 7.3.0]
Numpy 1.16.2
astropy 3.2.3
Scipy 1.2.1

Darwin-18.7.0-x86_64-i386-64bit
Python 3.7.6 (default, Jan  8 2020, 13:42:34) 
[Clang 4.0.1 (tags/RELEASE_401/final)]
Numpy 1.18.1
astropy 4.0
Scipy 1.4.1

**Comment by r-owen** [comment]
The current behavior might be an attempt to what some computer time-of-day clocks do on the day of a leap second. My understanding is that some operating systems gradually change that clock on the day of a leap seconds and others jump in some way. But I have not researched it in detail, nor have a thought of a way to test it. I think one would need to be running ntp or ptp in order for the system to see the leap second, but one would also need to be on the day of a leap second. That seems a tricky combination.

In any case UTC and TAI are well defined and always differ by an integer number of seconds. So the current behavior is incorrect.

**Comment by pllim** [comment]
FWIW, I still see the same behavior in 4.1.dev

**Comment by taldcroft** [comment]
@r-owen - the function you want is:
```
>>> def tai_minus_utc(ap_time):
...     time_iso = ap_time.iso  # ISO string representation of time
...     dt = Time(time_iso, scale='utc') - Time(time_iso, scale='tai')
...     return dt.sec
...
>>> ap_time = astropy.time.Time("2016-12-31 12:00:00", scale="utc", format="iso")
>>> tai_minus_utc(ap_time)
35.999999999996035
```

The issue you are seeing with `mjd` is not a bug in astropy, that just reflects that 1.0 JD on a leap-second day is one TAI second longer than 1.0 JD on a regular day.  This is also the fundamental way that SOFA / ERFA handles leap seconds.  1.0 JD is always the time from 00:00am one day to 00:00am the next day, and thus can be 86400 or 86401 seconds, depending.  The fractional JD goes from 0.0 to 1.0 linearly during that interval from 00:00 to the next 00:00.

**Comment by r-owen** [comment]
Yikes. Thank you for the explanation. I tried your function, and it works, but the result is:

  /opt/lsst/software/stack/python/miniconda3-4.7.10/envs/lsst-scipipe-4d7b902/lib/python3.7/site-packages/astropy/_erfa/core.py:107: ErfaWarning: ERFA function "dtf2d" yielded 1 of "time is after end of day (Note 5)"
    warnings.warn('ERFA function "' + func_name + '" yielded ' + wmsg, ErfaWarning)

I'll stick to my uglier code for now (actually I'm trying to convert utc to tai; but a function that computes the difference was simpler to demonstrate the issue).

**Comment by taldcroft** [comment]
That warning is annoying but benign.  It turns out that ERFA is more permissive than ISO and allows seconds to exceed 60.

**Comment by taldcroft** [comment]
But at a more fundamental level, your issue seems to be converting a time from UTC to TAI.  That is done with the code below and that conversion is always correct:
```
# 36 seconds between UTC and TAI
>>> t1_utc = astropy.time.Time("2016-12-31 12:00:00", scale="utc")
>>> t1_utc.tai
<Time object: scale='tai' format='iso' value=2016-12-31 12:00:36.000>

# Still 36 seconds in the middle of a leap second
>>> t2_utc = astropy.time.Time("2016-12-31 23:59:60.5", scale="utc")
>>> t2_utc.tai
<Time object: scale='tai' format='iso' value=2017-01-01 00:00:36.500>

# Now 37 seconds difference after the leap second
>>> t3_utc = astropy.time.Time("2017-01-01 00:00:00.5", scale="utc")
>>> t3_utc.tai
<Time object: scale='tai' format='iso' value=2017-01-01 00:00:37.500>
```

Now the key thing here is that the 37 seconds is the difference in the ISO representation of the same instant of time in two scales.  The function I showed is the converse, which is the physical time difference between the same ISO representation in two scales.

**Comment by r-owen** [comment]
Thank you. Basically I'm dealing with getting into and out of TAI in unix seconds (e.g. as returned by the CLOCK_TAI on linux systems) because that is our standard. In particular:
* Convert unix seconds to TAI unix seconds. This is not well defined, since different computers do different things with unix seconds during the day of a leap second. We avoid unix seconds when we can, but we have some cases where we are stuck with it. For example our control system uses DDS (a message passing system) and DDS tags its samples with unix seconds. We convert that to our standard.
* Convert other times to TAI unix seconds. This is usually well defined. I represent the times with astropy.time.Time, then extract TAI as MJD and do the obvious conversion manually (since astropy refuses to deal directly with TAI seconds).
* Convert TAI unix seconds to astropy time. A rare need, but we have at least one system that wants TAI as an ISO time. I convert TAI seconds to MJD and then use astropy.

**Comment by taldcroft** [comment]
Thanks for the explanation of your needs, that helps. 

From that perspective the trivial subclass that @mhvk showed is what you want for representing TAI unix seconds (physical seconds since the epoch).  From a naming perspective I would suggest calling this something like `TimeClockTai` or `TimeUnixClockTai` because for most people "Unix time" implies the Wikipedia definition (without leap seconds and independent of scale).  You can define this subclass in your own tools and it gets registered, so no need to have this in astropy core.

This subclass will fully and correctly handle the 2nd and 3rd bullet points.

The first bullet point, convert unix seconds to TAI  unix seconds is easy to do "approximately" and impossible to do perfectly all the time.  The "approximately" solution means that if the requirement on time tags accuracy (i.e. the time when DDS messages occur) is within 1 second, then the existing astropy `TimeUnix` class is sufficient.  

If tagging to higher accuracy is required at all times, then that is simply not possible with unix time because during a leap second the time tags are ambiguous and you can be off by up to 1.0 sec.  I.e. you get a repeat of times within the leap second and the single float value of "unix time" is not unique. This has the bad property that if you sort the DDS messages by unix time tag they can end up out of order.  If you don't care then you're done, if you do care then you have a real problem because the information in the DDS message is not sufficient.

**Comment by r-owen** [comment]
Thank you. I agree that it is pretty much impossible to figure out what unix time means to better than a second on the day of a leap second. We'll have to live with DDS received timestamps being a bit wonky on the day of a leap second. (Fortunately we mostly care about sent timestamps, and our code provides those as TAI).

I admit I am leery of using the `TimeUnix` class because it seems too magic. If my code is loaded then it appears, but it's not clear to users where it came from, making it difficult to document and debug.

Is my mjd-based solution markedly inferior in some way?

**Comment by taldcroft** [comment]
> I admit I am leery of using the TimeUnix class because it seems too magic. If my code is loaded then it appears, but it's not clear to users where it came from, making it difficult to document and debug.

This is your call.  We have tried to make this capability for user- or project-defined subclasses be a useful option.  For instance about the magic (not that everybody reads the documentation), in https://astropy.readthedocs.io/en/latest/time/index.html#time-format it shows right away how to see the available time format classes.  So for instance:
```
In [12]: Time.FORMATS                                                                                                                              
Out[12]: 
OrderedDict([('jd', astropy.time.formats.TimeJD),
             ('mjd', astropy.time.formats.TimeMJD),
             ('decimalyear', astropy.time.formats.TimeDecimalYear),
             ('unix', astropy.time.formats.TimeUnix),

 ..  Here your custom time format would appear with a different package root 
  for the class giving clear provenance and a way to get the documentation.
```

Using the customization machinery in the intended way means you get a fully functioning `Time` object with a clearly defined `format='unix_clock_tai'` `and all the normal behavior like getting the ISO representation of the time in TAI scale.  I do prefer this over explicitly computing a delta seconds between UTC and TAI, but that could be because I have an obvious bias in the way I handle times!

About your specific function to compute TAI minus UTC using MJD, this is just fine as far as I can see.

**Comment by mhvk** [comment]
Just as another pointer to user classes being used: `PINT` has been using their own `PulsarMJD` time format to deal with differences in leap seconds (indeed, it might be directly relevant - https://github.com/nanograv/PINT/blob/master/src/pint/pulsar_mjd.py). And in my own baseband package, I use a user format to deal with strange string storage of times at GMRT (see  https://github.com/mhvk/baseband/blob/767db98b22ea2e35bdc2eb3e25b3503b21588330/baseband/gsb/header.py#L20).

**Comment by r-owen** [comment]
Note that the constant `time.CLOCK_TAI` will be part of Python 3.9, making TAI in unix seconds readily available on Linux. That is an argument for adding standard support for TAI seconds to astropy.time: https://github.com/astropy/astropy/issues/10052

**Comment by taldcroft** [comment]
@r-owen - interesting about `time.CLOCK_TAI`.  Do you have a pointer to info about that?  (A quick google didn't turn up anything useful).  The general problem of getting the unique and correct current time on the local host is a useful thing to address, e.g. to ensure `Time.now()` is correct.

I'm just curious about the `time.CLOCK_TAI` implementation because of course that value is not actually a constant, e.g. you start a program running just before a leap second and the leap second offset changes during execution.  And for instance our Chandra operations computers currently run a very old linux (heavily firewalled) so no clue what sort of system services are required.  And for much of the world that runs on Windows, what is the support?

**Comment by mhvk** [comment]
@taldcroft - I asked nearly the same over at #10052, and got pointers (https://github.com/astropy/astropy/issues/10052#issuecomment-604069034 and following). @r-owen was the one who added it to python; it is linux only.  Maybe we should move discussion over there, since this is about possible support for it in `Time`?

**Comment by taldcroft** [comment]
@mhvk - yes it would be best to have the discussion in one thread.  Thinking more I'm on board with adding a `TimeUnixTAI` class.  I don't think it will rely on `time.CLOCK_TAI` since it is not needed in astropy Time.  Astropy Time already has built-in knowledge of leap seconds which works on all platforms, and I see the utility in having something that provides the linux-only TAI clock time to all platforms.   Chandra `cxcsec` is useful for good reason.

Looking at our implementation of `Time.now()` I think that calling `Time.now().unix_tai` during a leap second will give the wrong answer, but at all other times should match the linux CLOCK_TAI time.  I don't see a way around that but other than that everything should work fine.

**Comment by mhvk** [comment]
Possibly, `Time.now()` should get an argument which on linux systems allows one to get the right time. But agreed that this is somewhat separate from whether to have a `unix_tai` at all.

**Comment by r-owen** [comment]
Having a way to get current time from CLOCK_TAI would be nice, but we can certainly live without it, and it would only work on linux.

**Comment by taldcroft** [comment]
How does one actually get CLOCK_TAI within Python on a linux system (preferably without requiring C code)?

**Comment by r-owen** [comment]
In Python 3.9 the TAI time in unix seconds will be available as follows:
```
time.clock_gettime(time.time.CLOCK_TAI)
```
CLOCK_TAI is just another clock ID constant:  https://docs.python.org/3/library/time.html#time-clock-id-constants

**Comment by taldcroft** [comment]
@r-owen - sorry, I was asking how we can do this **now** for Python 3.6+, which is the current supported version requirement for astropy.  I was hoping for some linux built-in command to call from subprocess or some-such.

**Comment by taldcroft** [comment]
@r-owen - on this issue are there any outstanding items from your perspective or can we close this?

**Comment by r-owen** [comment]
@taldcroft I know of no direct way to get CLOCK_TAI from Python before 3.9. The documentation for the other CLOCK_x time constants suggests none were added during bug fix releases, so I did not push to add the feature to 3.8.x and earlier.

One can read CLOCK_TAI with a trivial C extension. I planned to write that before I thought to try to get it added to Python.

For now I use AstroPy.time to get TAI from time.time() in a way that works on days that have no leap second. I plan to do that until we switch to Python 3.9, unless a leap second looks likely to occur before we can upgrade to that version of Python.

Note that it is quite common for Linux operating systems to have no leap second table. In that situation CLOCK_TAI silently returns the same thing as time.time() and one has to decide whether to try to detect and handle that. (A scarier issue, but one I suspect is much rarer, is that the operating system has a stale leap second table.)

I agree that it is probably best for AstroPy.time to not try to read CLOCK_TAI. It is only available in Linux, is often wrong, and won't be available until Python 3.9 (unless you want to use a C extension).

Feel free to close this ticket. Thank you again for adding the TAI unix time scale!

**Comment by mhvk** [comment]
@r-owen - from 4.0 onward, astropy keeps track of leap seconds itself, so that would be another reason not to rely on `CLOCK_TAI` - astropy may well know better...

**Comment by r-owen** [comment]
It would be nice to be able to explicitly ask Astropy to use the leap second table maintained by a properly configured ntp or ptp, in order to avoid an unnecessary download, which can cause a surprising delay. But I agree that should not be the default. And that's for another issue.

**Comment by mhvk** [comment]
Right now, astropy already has the ability to look at a system file:  https://github.com/astropy/astropy/blob/60fb417789abd03269f4d3af412ee2004983e869/astropy/utils/iers/iers.py#L884-L888

where ` system_leap_second_file` is a configuration item: https://github.com/astropy/astropy/blob/60fb417789abd03269f4d3af412ee2004983e869/astropy/utils/iers/iers.py#L126-L128

So, it should be possible to configure that already!

p.s. I had intended to try to make sure that this is auto-configured depending on the system one is on, but I haven't gotten around to it yet.  It definitely could use more documentation... Am happy to help with that - though we should discuss that in a separate issue!
